### PR TITLE
Add Sod tube example as a nosetest case

### PR DESCRIPTION
**Pull request by tai271828**
Add a Sod tube nosetest example

This nosetest will provide a Sod tube simulation data and compare the data with 1D analytic solutions.

This pull request provides a nosetest based on the previous sandbox [Sod tube driving script] (https://github.com/solvcon/solvcon/blob/master/sandbox/gas/tube/go) with some different parameters. That is, this nosetest will run its own case, which differs from the sandbox driving script including

* This nosetest case removes unnecessary hooks (e.g. CFL hook) and reserves the hooks to generate data.
* This nosetest case use different edgelength, time_increment and steps_run to get higher CFL.
* This nosetest case is not launched over the sc command interface, but an explicit case registered to the arrangements.

To validate the SOLVCON data, this nosetest uses 1D anlytic solution as benchmark. It compares the numeric data with the following two algorithms. Both of them are de facto L2 norm method. Their names indicates different ways to exclude the data around the discontinuity:

* The so-called **counter** method
* The so-called **snapshot** method

# Counter Method

This method calculates the L2 norm of all given numerical data at every given time.

The regions adjacent discontinuities are likely to deviate much from analytic solutions. However, these data should be match the analytic solutions soon after few time iteration steps. The so-called counter comparison method does not regard the comparison as failure unless the data deviates from the analytic solutions many times in the given time period.

# Snapshot method

The snapshot method watches the final status of the numeric solutions and compare them to the analytic solutions. However, the snapshot method only picks up some of the locations which are not nearby the discontinuity region to watch, so it could avoid to compare the discontinuities.

# File Hierarchy

The commit https://github.com/tai271828/solvcon/commit/0a5b1e7389c5f572a00168d650f443d82f315792 introduces this new test example. It also shows the file hierarchy. In ``solvcon/examples/gas``

```
tests/
├── __init__.py
├── test_sod.py
└── tube
    ├── analytic.py
    ├── __init__.py
    └── sod.py
```

The other commits of this PR are auxiliary change for this "core" commit. 

# Test This PR

```
$ nosetests -v examples/gas/tests/
test_sod_tube_by_counter (tests.test_sod.TestGas) ... ok
test_sod_tube_by_snapshot (tests.test_sod.TestGas) ... ok

----------------------------------------------------------------------
Ran 2 tests in 7.721s

OK
```

# Topics Need to Discuss
## Add scipy dependency

The analytic solution uses the scipy module. Is it ok to install a module that is used by one application?

## Close Issue #128 

Issue #128 is ready to close. We can keep improving this Sod example in the future.

### CODE REVIEW

File: examples/gas/tests/tube/sod.py [hunk 1]
```diff
@@ -0,0 +1,379 @@
...
+        self.info('- rho = %.3f \n' % rho)
+        self.info('- v = %.3f \n' % v)
+        self.info('- p = %.3f \n' % p)
+
+
+class SodtubeMesher(object):
+    """
+    Representation of a Sod tube and the Gmsh meshing helper.
+
+    :ivar
```
Comment (yungyuc) [inline examples/gas/tests/tube/sod.py:121]: Redundant fields.

**Comment by yungyuc** [comment]
I am not aware of the "so-call" counter and snapshot methods.  Could you provide a reference for them?

For estimating error, a common practice is to use norm.  Set v(x, t) as the analytical solution and u(x, t) the numerical, the error field is e(x,t) = u(x,t) - v(x,t).  At any given time the norm of the error function can be calculated as L(t) = |e(x,t)|.  L_0, L_1, and L_2 are commonly used norms.

If we want to exclude the error around the discontinuity, those values around it may be exempted from the error field.  Thus we get the modified error function e'(x,t) = E(u(x,t) - v(x,t)) where E excludes the values around discontinuity.

And you probably need a report to show that your code really generates correct results.

File: examples/gas/tests/test_sod.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+from unittest import TestCase
+# import driving script
+from .tube import sod
+
+class TestGas(TestCase):
+
+    sod.run()
```
Comment (yungyuc) [inline examples/gas/tests/test_sod.py:37]: Why run this at the class level?  `sod.run()` doesn't return anything or set global variables.

File: solvcon/parcel/gas/probe.py [hunk 3]
```diff
@@ -149,9 +149,12 @@ def postmarch(self):
         return True
 
     def postloop(self):
+        #format_character = "%0" + str(len(self.points)) + "s"
+        format_character = "%03d"
         for point in self.points:
+            point_name = format_character % int(point.name)
```
Comment (yungyuc) [inline solvcon/parcel/gas/probe.py:155]: This looks like a bad idea since point.name isn't necessarily an integer.

**Review by yungyuc — CHANGES_REQUESTED** [review]

**Comment by tai271828** [comment]
The terms **counter method** and **snapshot method** are invented by myself. They are de facto L2 norm of the error function. Their names indicate that they adapt different algorithms to exclude the error around the discontinuity.

Your comment inspires me to implement one more intuitive and easier understanding test algorithm: use the analytic solution to get the coordinates of the discontinuity which is moving when time goes by, and then exclude the data around them.

Regarding the report, I will only say how close it is to the analytic solution, but not a "correct" result. For example, I noted this driving script gives npy binary files with different floating point numbers. I will write a report and check a bit why the npy files are not identical (by investigating a question like "does the deviation come from the discontinuity?") What do you think if this report just describe how close the numeric solution is to the analytic solution? Do you think it is enough so far?

**Comment by yungyuc** [comment]
Regarding the report, something shows your results in graphics will be good enough.  It's quite difficult to understand what the code does without graphics.

**Pull request closed without merging**